x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java

```diff
@@ -0,0 +1,108 @@
+                return;
+            }
+            final List<Tuple<PrivateKey, X509Certificate>> transportKeysAndCertificates =
+                ((StoreKeyConfig) transportKeyConfig).getPrivateKeyEntries(environment);
+            final List<Tuple<PrivateKey, X509Certificate>> httpCaKeysAndCertificates =
+                ((StoreKeyConfig) httpKeyConfig).getPrivateKeyEntries(environment).stream()
+                    .filter(t -> t.v2().getBasicConstraints() != -1).collect(Collectors.toList());
+            if (transportKeysAndCertificates.isEmpty()) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to enroll node. Elasticsearch node transport layer SSL configuration doesn't contain any keys"));
+                return;
+            } else if (transportKeysAndCertificates.size() > 1) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to enroll node. Elasticsearch node transport layer SSL configuration contains multiple keys"));
+                return;
+            }
+            if (httpCaKeysAndCertificates.isEmpty()) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration Keystore doesn't contain any " +
+                        "PrivateKey entries where the associated certificate is a CA certificate"));
+                return;
+            } else if (httpCaKeysAndCertificates.size() > 1) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration Keystore contain multiple " +
+                        "PrivateKey entries where the associated certificate is a CA certificate"));
+                return;
+            }
+
+            final String httpCaKey = Base64.getUrlEncoder().encodeToString(httpCaKeysAndCertificates.get(0).v1().getEncoded());
+            final String httpCaCert = Base64.getUrlEncoder().encodeToString(httpCaKeysAndCertificates.get(0).v2().getEncoded());
+            final String transportKey = Base64.getUrlEncoder().encodeToString(transportKeysAndCertificates.get(0).v1().getEncoded());
+            final String transportCert = Base64.getUrlEncoder().encodeToString(transportKeysAndCertificates.get(0).v2().getEncoded());
+            final List<String> nodeList = new ArrayList<>();
+            for (DiscoveryNode node : clusterService.state().getNodes()) {
+                nodeList.add(node.getAddress().toString());
+            }
+            listener.onResponse(new NodeEnrollmentResponse(httpCaKey,
+                httpCaCert,
+                transportKey,
+                transportCert,
+                clusterService.getClusterName().value(),
+                nodeList));
+
+        } catch (Exception e) {
```

Comment:
Nit: this is probably one of the few places where you don't have to worry about throwing exceptions if the method takes an action listener.

Reply:
Can you elaborate ? The idea here was to catch any Exceptions that might get thrown so that we call the action listener. On second thought though now, it's only the `getEncoded()` that would throw in the code above so I can have more contained try,catch and return a useful error with the exception

Reply:
I understand the intention. I meant to say that you don't have to do it because it's already handled this way in https://github.com/elastic/elasticsearch/blob/668a72dec5dc06a56ba5753e093c1731143d5941/server/src/main/java/org/elasticsearch/action/support/TransportAction.java#L79 

In general, I'm not a fan of expansive try/catch blocks as the intention behind the catch is lost, it feels more like "ignore exceptions".

This is just a nit.

Reply:
Thanks Albert, makes sense. I'll limit the try block scope and have any unexpected ones handled by the base class.